RTC left panel

martinRenou opened the pull request:
## References

Rework some parts of https://github.com/jupyterlab/jupyterlab/pull/11180, somewhat related to https://github.com/jupyterlab/jupyterlab/pull/11676

cc. @hbcarlos @SylvainCorlay @wolfv 

## Code changes

Add two new plugins in the user-extension: 

- rtcGlobalAwareness: adds a global YJS awareness for tracking who is connected and who is working on which Notebook/file
- rtcLeftPanel: adds a left panel showing the current user and who is connected 

## User-facing changes

![collaborators](https://user-images.githubusercontent.com/21197331/154983959-5bd44e92-b07d-4ab7-aa82-3eb0ddaf2a3f.png)

## Backwards-incompatible changes

None

martinRenou commented:
Remaining:

- [ ] : Add logic for clicking on collaborators name and show the Notebook/file they are working on

Open questions:

- Should we rename the user-extension? Should this be in a another extension (rtc-extension)?
- Can we/should we make some components/plugins reusable for https://github.com/jupyterlab/jupyterlab/pull/11676 ?

martinRenou commented on packages/user/src/tokens.ts:
```diff
@@ -102,7 +117,7 @@ export namespace IUser {
      * User's cursor color and icon color if avatar_url is undefined
      * (there is no image).
      */
-    readonly color: string;
```
Should be reverted

jtpio commented:
Thanks Martin for starting this :+1: 

> Should we rename the user-extension? Should this be in a another extension (rtc-extension)?

Probably yes. Especially if we start adding more rtc features that are not necessarily related to a "user". So `rtc-extension` or maybe `collaboration-extension` would work (other ideas welcome).

martinRenou commented:
> Probably yes. Especially if we start adding more rtc features that are not necessarily related to a "user". So rtc-extension or maybe collaboration-extension would work (other ideas welcome).

Thanks! I'll rename it to `collaboration-extension` if nobody is against it.

martinRenou commented:
Implemented the click logic for going to the collaborator's location 

https://user-images.githubusercontent.com/21197331/155149827-54df5790-cef6-4ec7-bc8b-9dbc410d8592.mp4

martinRenou commented on docs/source/extension/extension_points.rst:
```diff
@@ -125,9 +125,9 @@ might want to use the services in your extensions.
   Use this if you want to access to the identity of the current connected user.
-- ``@jupyterlab/user:IUserMenu``: A service for the user menu on the application.
+- ``@jupyterlab/collaboration:IUserMenu``: A service for the user menu on the application.
   Use this if you want to add new items to the user menu.
```
TODO: Add extensions points here

martinRenou commented on packages/collaboration-extension/README.md:
```diff
@@ -1,3 +1,3 @@
-# @jupyterlab/user-extension
+# @jupyterlab/collaboration-extension
 
 A JupyterLab package which provides a set of plugins to represent an user.
```
Update this doc

martinRenou commented on packages/collaboration/README.md:
```diff
@@ -1,3 +1,3 @@
-# @jupyterlab/user
+# @jupyterlab/collaboration
 
 A JupyterLab package which provides a user set of widgets.
```
Update this doc

martinRenou commented on packages/collaboration/package.json:
```diff
@@ -1,5 +1,5 @@
-  "name": "@jupyterlab/user",
+  "name": "@jupyterlab/collaboration",
   "version": "4.0.0-alpha.4",
   "description": "JupyterLab - User Widget",
```
Update description

jtpio commented:
Some CI checks will be failing due to the rename of the `user-extension` package.

This means a new `4.0.0ax` release will be required after the merge so the packages are available on npm.

trungleduc commented:
> Implemented the click logic for going to the collaborator's location
> 
>  collaborators.mp4

It looks great @martinRenou! I think the collaborator's location should be limited to the opened files only and not all their activities.

jtpio commented:
> Thanks! I'll rename it to collaboration-extension if nobody is against it.

Sounds good, thanks. Maybe that's something we could also bring to the weekly meeting for consideration.

martinRenou commented:
> It looks great @martinRenou! I think the collaborator's location should be limited to the opened files only and not all their activities.

@trungleduc I agree! I'll do that and rebase my PR

jtpio commented on packages/collaboration-extension/src/index.ts:
```diff
@@ -0,0 +1,210 @@
+        awareness: awareness
+      }
+    );
+    websocket;
```
Looks like this line can be removed?

```suggestion
```

jtpio commented on packages/collaboration/src/tokens.ts:
```diff
@@ -29,7 +30,23 @@ export const ICurrentUser = new Token<ICurrentUser>(
+ * The user panel token.
+ */
+export const IUserPanel = new Token<AccordionPanel>(
+  '@jupyterlab/collaboration:userPanel'
```
For consistency we can rename it to:

```suggestion
  '@jupyterlab/collaboration:IUserPanel'
```

jtpio commented on packages/collaboration/src/tokens.ts:
```diff
@@ -29,7 +30,23 @@ export const ICurrentUser = new Token<ICurrentUser>(
+ * The global awareness token.
+ */
+export const IGlobalAwareness = new Token<Awareness>(
+  '@jupyterlab/collaboration:rtcGlobalAwareness'
```
Another minor tweak for consistency:

```suggestion
  '@jupyterlab/collaboration:IGlobalAwareness'
```

jtpio commented:
Looks like there are a couple of conflicts to fix as a follow-up to the recent changes and release.

martinRenou commented:
Thanks for the review, I'll rebase fix conflicts and make your suggested changes

jtpio commented:
Thanks Martin :+1: 

cc @hbcarlos if you have any input on the packages being renamed to `@jupyterlab/collaboration` and `@jupyterlab/collaboration-extension`, since these were added in https://github.com/jupyterlab/jupyterlab/pull/11443.

hbcarlos commented:
I think is a great idea since there is a discussion open about moving the ICurrentUser into services #11678.

jtpio commented:
@martinRenou looks like the integrity check is giving the following error:

![image](https://user-images.githubusercontent.com/591645/158192362-836eb473-5bff-479a-988e-a2bf8ad4fd25.png)

martinRenou commented:
Thanks Jeremy, this should be addressed now

jtpio commented:
Looks like the UI test failures are not related (we could double check on other PRs to see if that's the case there too):

![image](https://user-images.githubusercontent.com/591645/158213289-64344891-3f23-4f0b-9437-05853a729c7b.png)

jtpio commented:
> we could double check on other PRs to see if that's the case there too

They seem to have passed in https://github.com/jupyterlab/jupyterlab/pull/12215

jtpio commented:
@martinRenou looks like this needs a lint pass:

https://github.com/jupyterlab/jupyterlab/runs/5550144954?check_suite_focus=true

![image](https://user-images.githubusercontent.com/591645/158343194-4554ee84-ee89-40d5-a6db-740b05983cb8.png)

jtpio commented:
> looks like this needs a lint pass

Done in https://github.com/jupyterlab/jupyterlab/pull/12095/commits/ba7b963d6864165a1fe1eb4e562ea9cbba16dbaf

jtpio commented on packages/collaboration-extension/src/index.ts:
```diff
@@ -0,0 +1,201 @@
+/**
+ * Jupyter plugin creating a global awareness for RTC.
+ */
+const rtcGlobalAwarenessPlugin: JupyterFrontEndPlugin<Awareness | null> = {
```
Not sure this plugin should provide a `Awareness | null`. Other plugins requiring `IGlobalAwareness` will expect the service to be available. It might be a bit odd if they have to check whether it is `null` before using it.

martinRenou commented:
Kicking CI

martinRenou commented:
@jtpio I rebased and took your comment into account and implemented a mock for when the collaborative flag is not set.

jtpio commented:
I just published `@jupyterlab/collaboration@4.0.0-alpha.11` to `npm`: https://www.npmjs.com/package/@jupyterlab/collaboration

This should help make most of the CI pass now.

jtpio commented on dev_mode/package.json:
```diff
@@ -183,19 +183,22 @@
     "chokidar": "^3.4.0",
     "css-loader": "^6.7.1",
     "duplicate-package-checker-webpack-plugin": "^3.0.0",
+    "file-loader": "~6.0.0",
```
Probably the reason for some of the CI failures.

jtpio commented on jupyterlab/staging/package.json:
```diff
@@ -183,19 +183,22 @@
     "chokidar": "^3.4.0",
     "css-loader": "^6.7.1",
     "duplicate-package-checker-webpack-plugin": "^3.0.0",
+    "file-loader": "~6.0.0",
```
Similar to the comment above, looks like these should be removed.

jtpio commented:
Looks like there are a couple more left overs. Will push an update to the branch.

martinRenou commented:
Thank you for pushing on this!

jtpio commented:
From https://github.com/jupyterlab/jupyterlab/runs/7498086273?check_suite_focus=true:

```
Parsed an error from text content: undefined Error: No provider for: @jupyterlab/user:ICurrentUser.
```

This is likely because `jupyterlab/staging/yarn.lock` is referencing the published `@jupyterlab/user` package which exposes `@jupyterlab/user:ICurrentUser`, while the token has now been renamed to `@jupyterlab/collaboration:ICurrentUser`.

Which can also be seen locally when running `jupyter lab --core-mode`:

![image](https://user-images.githubusercontent.com/591645/180767212-7ef2bdaf-1562-4cea-92e5-e9bd12130e54.png)

Normally this should be fixed once we make a new release and the `alpha.12` packages start depending on `@jupyterlab/collaboration`.

jtpio commented:
Also seeing these locally:

```
[E 2022-07-25 13:33:38.259 ServerApp] Uncaught exception GET /api/yjs/JupyterLab:globalAwareness (::1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='GET', uri='/api/yjs/JupyterLab:globalAwareness', version='HTTP/1.1', remote_ip='::1')
    Traceback (most recent call last):
      File "/home/user/micromamba/envs/jupyterlab-tmp/lib/python3.10/site-packages/tornado/websocket.py", line 944, in _accept_connection
        await open_result
      File "/home/user/git/userio/jupyterlab-tmp/jupyterlab/handlers/ydoc_handler.py", line 76, in open
        self.room = self.websocket_server.get_room(path)
      File "/home/user/git/userio/jupyterlab-tmp/jupyterlab/handlers/ydoc_handler.py", line 32, in get_room
        file_format, file_type, file_path = path.split(":", 2)
    ValueError: not enough values to unpack (expected 3, got 2)
```

jtpio commented on packages/collaboration-extension/src/index.ts:
```diff
@@ -0,0 +1,204 @@
+    const server = ServerConnection.makeSettings();
+    const url = URLExt.join(server.wsUrl, 'api/yjs');
+
+    new WebsocketProvider(url, 'JupyterLab:globalAwareness', ydoc, {
```
Looks like the handler exposed by `jupyter-server-ydoc` looks for a path: 

https://github.com/jupyter-server/jupyter_server_ydoc/blob/deb54d2b036bf02911367e9c9b38435c1966df65/jupyter_server_ydoc/ydoc.py#L33

jtpio replied:
cc @davidbrochart who might have some input on this.

I think the original idea was to reuse the same endpoint for the global awareness. Although in that case there is no file. Wondering whether passing a placeholder would be enough, or maybe use a different approach.

jimgoo replied:
FWIW, with this rtc_left_panel branch merged with current master (4273278a5e3c174ba7469ef40f3e5cae43339376), I'm seeing an error in the referenced line of `jupyter-server-ydoc` when `path=JupyterLab:globalAwareness`:

```
[E 2022-07-29 13:35:53.436 ServerApp] Uncaught exception GET /api/yjs/JupyterLab:globalAwareness (::1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='GET', uri='/api/yjs/JupyterLab:globalAwareness', version='HTTP/1.1', remote_ip='::1')
    Traceback (most recent call last):
      File "/Users/jimgoo/miniconda3/envs/jupyterlab/lib/python3.8/site-packages/tornado/websocket.py", line 944, in _accept_connection
        await open_result
      File "/Users/jimgoo/git/jupyter_server_ydoc/jupyter_server_ydoc/ydoc.py", line 119, in open
        self.room = self.websocket_server.get_room(path)
      File "/Users/jimgoo/git/jupyter_server_ydoc/jupyter_server_ydoc/ydoc.py", line 60, in get_room
        file_format, file_type, file_path = path.split(":", 2)
    ValueError: not enough values to unpack (expected 3, got 2)
```

davidbrochart replied:
Sorry, I missed that one @jtpio.
I need to see with @martinRenou, but I think we can make a special case for the global awareness, where the endpoint would just be `/api/yjs/globalAwareness`. Or maybe we can fake the file format and type with `/api/yjs/foo:bar:globalAwareness`. We basically want a Y room just for the awareness.

jimgoo replied:
@davidbrochart I managed to fake something out via:
```typescript
new WebsocketProvider(url, 'foo:bar:globalAwareness', ydoc, {
      awareness: awareness
});
```

Inside of `jupyter_server_ydoc` (https://github.com/jupyter-server/jupyter_server_ydoc/blob/main/jupyter_server_ydoc/ydoc.py#L59), this gives

```
ystore_class: <class 'jupyter_server_ydoc.ydoc.JupyterSQLiteYStore'>
path:  foo:bar:globalAwareness 
file_format: foo, file_type: bar, file_path: globalAwareness
updates_file_path: .bar:globalAwareness.y
```

At first I was seeing a lot of messages like `404 GET /api/yjs/foo:bar:globalAwareness (::1): file or directory does not exist: 'globalAwareness'`, but I just faked that out by doing:
```
touch globalAwareness
```

That seems to work fine as far as getting this branch merged with master working. I only notice the occasional messages like these, which I don't think are related to the faking out of globalAwareness: 
```
[I 2022-07-29 15:56:59.957 ServerApp] Creating new notebook in
[I 2022-07-29 15:57:00.224 ServerApp] Kernel started: 1f9f65aa-5eca-4337-8583-937f494129ba
Exception in callback <TaskWakeupMethWrapper object at 0x119b19a90>(<Future finished result=None>)
handle: <Handle <TaskWakeupMethWrapper object at 0x119b19a90>(<Future finished result=None>)>
Traceback (most recent call last):
  File "/Users/jimgoo/miniconda3/envs/jupyterlab/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
RuntimeError: Cannot enter into task <Task pending name='Task-859' coro=<WebsocketServer.serve() running at /Users/jimgoo/miniconda3/envs/jupyterlab/lib/python3.8/site-packages/ypy_websocket/websocket_server.py:103> wait_for=<Future finished result=None>> while another task <Task pending name='Task-96793' coro=<PeriodicCallback._run() running at /Users/jimgoo/miniconda3/envs/jupyterlab/lib/python3.8/site-packages/tornado/ioloop.py:921> cb=[IOLoop.add_future.<locals>.<lambda>() at /Users/jimgoo/miniconda3/envs/jupyterlab/lib/python3.8/site-packages/tornado/ioloop.py:694]> is being executed.
Task was destroyed but it is pending!
task: <Task pending name='Task-859' coro=<WebsocketServer.serve() running at /Users/jimgoo/miniconda3/envs/jupyterlab/lib/python3.8/site-packages/ypy_websocket/websocket_server.py:103> wait_for=<Future finished result=None>>```

The RTC left panel is amazing and works as advertised @martinRenou!

davidbrochart replied:
Thanks for checking @jimgoo.
What you are seeing is jupyter-server-ydoc trying to save Y updates into a YStore, but I don't think this makes sense for the awareness. I'll update it to support rooms that don't have a file associated with them (feel free to open a PR if you feel like it).

> I only notice the occasional messages like these, which I don't think are related to the faking out of globalAwareness

I think this is a Tornado issue, I've been seeing that lately too.

davidbrochart replied:
I released jupyter-server-ydoc v0.1.10. It should work with this PR.

jimgoo replied:
Thanks for the help @davidbrochart, the new version of jupyter-server-ydoc is working great. 

@martinRenou I took what you did in this branch and extended it to sync most of StateDB whenever a collaborator is clicked (https://github.com/jimgoo/jupyterlab/tree/rtc-layout-sync). It's kind of in the spirit of VSCode's LiveShare "Follow" feature, which I found had bugs with notebook files. I still want to sync a collaborator's scroll position inside of each widget as soon as I figure out how.

https://user-images.githubusercontent.com/2735842/182041267-36ec1819-4c45-4754-a10b-e9a09fae0147.mov

martinRenou replied:
Sorry I missed the notifications here.

Thank you @davidbrochart for the fix!

@jimgoo that's great! Would you like to push a PR against my branch to include your commits here?

davidbrochart commented:
Thanks for checking @jimgoo.
What you are seeing is jupyter-server-ydoc trying to save Y updates into a YStore, but I don't think this makes sense for the awareness. I'll update it to support rooms that don't have a file associated with them (feel free to open a PR if you feel like it).

> I only notice the occasional messages like these, which I don't think are related to the faking out of globalAwareness

I think this is a Tornado issue, I've been seeing that lately too.

martinRenou commented:
@jtpio I rebased, squashed and fix the one small conflict there was.

Those tests will pass when `collaboration` is published, right?
```
[webpack-cli] ModuleNotFoundError: Module not found: Error: Can't resolve '@jupyterlab/user/style/index.js' in '/home/runner/work/jupyterlab/jupyterlab/jupyterlab/staging/node_modules/@jupyterlab/docprovider-extension/style'
```

jtpio commented:
Hmm strange I don't recall seeing these while looking at it in https://github.com/jupyterlab/jupyterlab/pull/12095#issuecomment-1193926552.

> rebased, squashed 

Did you push additional changes as part of the rebase / squash? (hard to see with the GitHub UI)

martinRenou commented:
> Hmm strange I don't recall seeing these while looking at it in https://github.com/jupyterlab/jupyterlab/pull/12095#issuecomment-1193926552.

I am not sure I understand, you are also saying in your comment that you see the same error. So we are on the same page? Or am I missing something?

martinRenou commented:
> Did you push additional changes as part of the rebase / squash? (hard to see with the GitHub UI)

Not that I'm aware of, though I re-generated the `yarn.lock` for which the conflict was too hard to fix manually. I also removed a trailing mention to `@jupyterlab/user` which I believe came from the rebase.

martinRenou commented:
> I am not sure I understand, you are also saying in your comment that you see the same error. So we are on the same page? Or am I missing something?

My bad, it's not really the same error. I was confused because I also saw `No provider for: @jupyterlab/user:ICurrentUser` somewhere else.

jtpio commented:
BTW feel free to push many commits to the branch as it might make it easier to troubleshoot this.

The general practice in the jupyterlab repo recently has been to "Squash and Merge" anyways.

fcollonval commented on packages/collaboration-extension/src/index.ts:
```diff
@@ -0,0 +1,204 @@
+      return;
+    }
+
+    const userPanel = new AccordionPanel({
+      renderer: new RTCPanel.Renderer()
+    });
```
There is now a `SidePanel` that contains a `AccordionPanel` with the proper renderer in `ui-components`:

```suggestion
    const userPanel = new SidePanel();
```

fcollonval commented on packages/collaboration/src/collaboratorspanel.tsx:
```diff
@@ -0,0 +1,148 @@
+
+    this._awareness = awareness;
+
+    this.title.label = 'Online Collaborators';
```
We need to translate that string - I would advice setting it in the activation function of the plugin rather than here to avoid bring down the translator object.

fcollonval commented on packages/collaboration/src/sidepanel.tsx:
```diff
@@ -0,0 +1,43 @@
+import { caretDownIcon } from '@jupyterlab/ui-components';
+import { AccordionPanel, Title, Widget } from '@lumino/widgets';
+
+/**
+ * RTC accordion panel customization
+ */
+export namespace RTCPanel {
+  /**
+   * Custom renderer for the RTC sidebar
+   */
+  export class Renderer extends AccordionPanel.Renderer {
+    /**
+     * Render the collapse indicator for a section title.
+     *
+     * @param data - The data to use for rendering the section title.
+     *
+     * @returns A element representing the collapse indicator.
+     */
+    createCollapseIcon(data: Title<Widget>): HTMLElement {
+      const iconDiv = document.createElement('div');
+      caretDownIcon.element({
+        container: iconDiv
+      });
+      return iconDiv;
+    }
+
+    /**
+     * Render the element for a section title.
+     *
+     * @param data - The data to use for rendering the section title.
+     *
+     * @returns A element representing the section title.
+     */
+    createSectionTitle(data: Title<Widget>): HTMLElement {
+      const handle = super.createSectionTitle(data);
+      handle.classList.add('jp-AccordionPanel-title');
+      return handle;
+    }
+  }
+}
```
Unneeded any more - prefer using `SidePanel`.

```suggestion
```

fcollonval commented on packages/collaboration-extension/src/index.ts:
```diff
@@ -0,0 +1,204 @@
+  id: '@jupyterlab/collaboration-extension:rtcPanel',
+  autoStart: true,
+  requires: [ICurrentUser, IGlobalAwareness],
+  provides: IUserPanel,
```
We do not provide anything

```suggestion
```

fcollonval commented on packages/collaboration/src/tokens.ts:
```diff
@@ -29,7 +30,28 @@ export const ICurrentUser = new Token<ICurrentUser>(
+  '@jupyterlab/collaboration:IUserMenu'
+);
+
+/**
+ * The user panel token.
+ */
+export const IUserPanel = new Token<AccordionPanel>(
+  '@jupyterlab/collaboration:IUserPanel'
+);
+
```
We do not provide such token (and we should not provide a widget as token).

```suggestion
```

fcollonval commented on packages/collaboration/src/userinfopanel.tsx:
```diff
@@ -0,0 +1,64 @@
+
+  constructor(user: ICurrentUser) {
+    super({});
+    this.title.label = 'User info';
+    this.title.caption = 'User information';
```
This requires translation - I would advice to do it in the plugin activation to avoid bringing down the translator object.

fcollonval commented on packages/collaboration/style/sidepanel.css:
```diff
@@ -0,0 +1,135 @@
+  font-size: var(--jp-ui-font-size1);
+}
+
+.jp-RTCPanel .lm-AccordionPanel-title {
+  box-sizing: border-box;
+  background: #e5e5e500;
+  font: 12px Helvetica, Arial, sans-serif;
+  min-height: 22px;
+  max-height: 22px;
+  min-width: 35px;
+  line-height: 25px;
+  margin: 0;
+  display: flex;
+  background-color: var(--jp-layout-color2);
+  border-bottom: solid var(--jp-border-width) var(--jp-border-color1);
+  border-top: solid var(--jp-border-width) var(--jp-border-color1);
+  box-shadow: var(--jp-toolbar-box-shadow);
+}
+
```
Should not be needed to have homogeneous styling with others side panels.

```suggestion
```

fcollonval reviewed (changes requested):
I made some suggestions.

martinRenou commented:
> Which is why I was wondering about additional changes.

@jtpio is it because I changed stuffs under `staging`? Should I keep this directory untouched?

jtpio commented:
For that particular case it's ok if the staging check fails because of the rename of the packages.

fcollonval reviewed (approved):
Thanks for the changes @martinRenou

jtpio commented:
Nice. We'll still want to hold for a bit before merging since this requires making a new pre-release right after.
